File: osint_nexus/core/detectors/http.py
from collections.abc import Mapping
from typing import TypedDict

from beartype import beartype

from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue

HttpData = Mapping[str, JSONValue]


class HttpResultData(TypedDict):
    platform: JSONValue
    mobile: JSONValue
    architecture: JSONValue
    language: JSONValue
    full_headers: Mapping[str, JSONValue]


class HttpResult(TypedDict):
    name: str
    data: HttpResultData
    confidence: float


class HttpFingerprintStrategy(FingerprintStrategy[HttpData, Mapping[str, JSONValue]]):
    """Strategy for parsing HTTP headers for device info."""

    name: str = "http_headers"

    @beartype
    def extract(self, data: HttpData) -> Mapping[str, JSONValue]:
        # More precise: ensure all header values are strings
        headers: dict[str, str] = {k: str(v) for k, v in data.items()}

        # Deep extraction of all Sec-CH-UA headers
        sec_ch_ua_headers = {k: v for k, v in data.items() if k.lower().startswith("sec-ch-ua")}

        fingerprint: HttpResultData = {
            "platform": headers.get("sec-ch-ua-platform"),
            "mobile": headers.get("sec-ch-ua-mobile") == "?1",
            "architecture": headers.get("sec-ch-ua-arch"),
            "language": headers.get("accept-language"),
            "full_headers": sec_ch_ua_headers,
        }

        return {
            "name": self.name,
            "data": fingerprint,
            "confidence": 0.85,
        }

File: internal/engine/strategies/http.go
package strategies

import (
	"context"
	"strings"

	"github.com/FJ-cyberzilla/osint-nexus/internal/types"
)

// HTTPPayload implements types.FingerprintPayload for HTTP fingerprinting input.
type HTTPPayload struct {
	Headers map[string]string `json:"headers"`
}

// PayloadType returns the payload type identifier.
func (p HTTPPayload) PayloadType() string { return "http_input" }

// HTTPOutputPayload implements types.FingerprintPayload for HTTP fingerprinting output.
type HTTPOutputPayload struct {
	Platform     string            `json:"platform"`
	Mobile       bool              `json:"mobile"`
	Architecture string            `json:"architecture"`
	Language     string            `json:"language"`
	FullHeaders  map[string]string `json:"full_headers"`
}

// PayloadType returns the payload type identifier.
func (p HTTPOutputPayload) PayloadType() string { return "http_output" }

// HttpFingerprintStrategy identifies device info based on HTTP headers.
type HttpFingerprintStrategy struct{}

// NewHttpFingerprintStrategy initializes a new HttpFingerprintStrategy.
func NewHttpFingerprintStrategy() *HttpFingerprintStrategy {
	return &HttpFingerprintStrategy{}
}

// Name returns the strategy identifier.
func (s *HttpFingerprintStrategy) Name() string {
	return "http_headers"
}

// Extract analyzes HTTP headers to extract device information.
func (s *HttpFingerprintStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(HTTPPayload)
	if !ok {
		return types.FingerprintResult{
			Name: s.Name(),
			Data: types.FingerprintData{
				Type:    "http_headers",
				Payload: HTTPOutputPayload{},
			},
			Confidence: 0.85,
		}, nil
	}

	headers := make(map[string]string)
	for k, v := range payload.Headers {
		headers[strings.ToLower(k)] = v
	}

	secChUaHeaders := make(map[string]string)
	for k, v := range payload.Headers {
		if strings.HasPrefix(strings.ToLower(k), "sec-ch-ua") {
			secChUaHeaders[k] = v
		}
	}

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type: "http_headers",
			Payload: HTTPOutputPayload{
				Platform:     headers["sec-ch-ua-platform"],
				Mobile:       headers["sec-ch-ua-mobile"] == "?1",
				Architecture: headers["sec-ch-ua-arch"],
				Language:     headers["accept-language"],
				FullHeaders:  secChUaHeaders,
			},
		},
		Confidence: 0.85,
	}, nil
}
